### PR TITLE
feat(HMS-1245): UI AWS permission check

### PR DESCRIPTION
**Pull request by avitova**
The best way to test this is to insert [a new permission](https://github.com/RHEnVision/provisioning-backend/blob/8bbd98e720e8063ca3bc86b01303af046e636962/internal/clients/http/ec2/iam_client.go#L39) and use your local backend.

### CODE REVIEW

**Comment by akhil-jha** [comment]
Env Build error. Running test again. 
/retest

**Comment by lzap** [comment]
Do you have a screenshot of the error message? :-)

**Comment by avitova** [comment]
After a conversation on Slack, we have decided to change this from a step in the stepper to a warning shown only in case of the error. Let me know what you think.
To reproduce this, you can for example use your sources on stage, as well as the image builder, but run the frontend with your local backend where you change the expectedStatement variable. Make sure to produce an error so that the warning is shown. Let me know what you think of this.
image

**Comment by lzap** [comment]
@amirfefer have another PR where he shows a warning message in a different format, please talk to each other and pick one style.

Nitpick: join the permissions with `, ` (comma space)

File: src/Components/ProvisioningWizard/steps/ReservationProgress/index.js [hunk 1]
```diff
@@ -50,6 +51,12 @@ const ReservationProgress = ({ setLaunchSuccess }) => {
   ] = useWizardContext();
   const { nextInterval, currentInterval } = useInterval(POLLING_BACKOFF_INTERVAL);
 
+  // check AWS permissions
+  const { data } = useQuery([`permissions`, `${chosenRegion}-${chosenSource}`], () => checkAWSPermissions(chosenSource, chosenRegion), {
```
Comment (ezr-ondrej) [inline src/Components/ProvisioningWizard/steps/ReservationProgress/index.js:55]: hmm I wonder... shall we run the check only if the provider is AWS?

**Review by ezr-ondrej — COMMENTED** [review]
Did we test this with Azure? The request for permissions will probably fail, right, do we handle it gracefully? The code doesn't seem to do this explicitly, so at minimum it reads bad, because we run `checkAWSPerms` for all providers.

**Just a thought** (feel free to ignore): did you think about checking the permissions reactively, maybe when we get AWS error. In that way we would save up a request to API and AWS, I do not think it's strictly necessary and this might be a better solution :)

**Comment by lzap** [comment]
> did you think about checking the permissions reactively

Yes, for AWS it means parsing string error and trying to figure out what was wrong. Not too clean.

But good point - this must not break Azure/GCP and it should be implemented in a way that we can easily add the same support into Azure/GCP too.

**Comment by lzap** [comment]
> because we run `checkAWSPerms` for all providers.

This is bad, needs to be fixed.

**Comment by amirfefer** [comment]
My concern is it might confuse users. AFAIK,  the error doesn't have to be related to the missing permission, but the error message is hidden, and the warning is highlighted, an advice from @MariSvirik  would help :)

File: src/Components/ProvisioningWizard/steps/ReservationProgress/index.js [hunk 2]
```diff
@@ -178,6 +185,15 @@ const ReservationProgress = ({ setLaunchSuccess }) => {
...
             <InstancesTable provider={provider} region={chosenRegion} reservationID={reservationID} />
           )}
+          {currentError && missingPermissions.length != 0 && (
+            <Alert
+              data-testid="insufficient_permissions_alert"
+              ouiaId="insufficient_permissions_alert"
+              variant="warning"
+              isInline
+              title={`Warning: Insufficient permissions for AWS account, following permissions missing: ${missingPermissions}.`}
+            />
```
Comment (lzap) [inline src/Components/ProvisioningWizard/steps/ReservationProgress/index.js:195]: So launch will work if the user is missing ListPermission or other related permissions which are only required for the check. Can you reword the warning to something like:

_Warning: Launch may fail, the following permissions are missing: xxxx_

**Comment by lzap** [comment]
I take my comments on the warning UI back, @amirfefer explained why it is difficult to implement warnings as we would like to. Let’s focus that on later, for now this is a good enough solution.

Summary from our discussion:

* It would be nice to have the check as an extra step before or after Create reservation.
* This step would indicate a warning (perhaps a clickable pop over).
* Multiple warning steps should be supported ideally for the future.
* Significant code changes are required to support that - not subject for this PR.

**Comment by MariSvirik** [comment]
If the permissions are missing launch will fail, correct?
Right now I see a yellow alert, and a red exclamation mark inside the step and a neutral large icon on the top...so I am not sure if this is still in progress or not.

@lzap what would be that extra step? Do you envision it living inside the wizard?

**Comment by avitova** [comment]
@MariSvirik No, they will not. Users could set their permissions according to our recommendations, and in that case, we'd be able to find out missing permissions accurately.
But there are ways in which they would not follow [our instructions](https://github.com/RHEnVision/provisioning-backend/blob/main/docs/configure-amazon-role.md), and their setup would work anyways. For example, instead of all the needed permissions, they just insert something like "iam:*" and "ec2:*", which would grant them all currently needed permissions. There are many reasons why it is impossible to detect these rules 100 % accurately.
TLDR; this is just a heads-up for users to check their permissions. I would say this would be helpful when we change the list of expected permissions and need users to update their AWS policies.

**Comment by lzap** [comment]
Now that I think about it, I think the alternative (and maybe better) solution to this is to take this check few screens before when a source is selected. After source is selected, we can trigger an event of requesting source permission check and display a warning there. What you all think? @ezr-ondrej @amirfefer @avitova @MariSvirik

**Comment by MariSvirik** [comment]
@lzap that would be great. 
Let's tell the users as soon as possible that there is a problem with permissions.
How long would it take to check?

**Comment by lzap** [comment]
Tens of milliseconds.

**Comment by avitova** [comment]
image
So I tried to implement what Maria proposed. Let me know what you think about it now.
We need to now the region and source to find out properly what permissions are missing. This should call the permission validation for every change in one of those fields. I thought it might be 1) safer and 2) better for the future to make the helper function more generic for any provider, so that we only return an empty array for other providers. Let me know your thoughts. :)

**Comment by MariSvirik** [comment]
@avitova How can users fix it? We should tell them.
To make it general we can still include an inline alert in this step with an expandable section that would include missing permissions, but that would be too prominent in my opinion.
Is there any other way to find which permissions are missing if we don't name them in the helper text?

I pinged the doc team to check the text.
Screenshot 2023-08-22 at 12 21 29

File: src/Components/ProvisioningWizard/steps/AccountCustomizations/aws.js [hunk 3]
```diff
@@ -54,6 +61,13 @@ const AccountCustomizationsAWS = ({ setStepValidated, image }) => {
             }))
           }
         />
+        {missingPermissions.length != 0 && (
+          <HelperText>
+            <HelperTextItem variant="warning">
+              Warning: Launch may fail, the following permissions are missing: {missingPermissions.join(', ')}
```
Comment (ragevou) [inline src/Components/ProvisioningWizard/steps/AccountCustomizations/aws.js:67]:
```suggestion
              Warning: Launch can fail, the following permissions are missing: {missingPermissions.join(', ')}
```

File: src/API/index.js [hunk 4]
```diff
@@ -72,3 +72,14 @@ export const fetchLaunchTemplates = async (sourceID, region) => {
...
 };
+
+export const checkPermissions = async (provider, sourceID, region) => {
+  switch (provider) {
+    case AWS_PROVIDER: {
+      const { data } = await axios.get(provisioningUrl(`sources/${sourceID}/validate_permissions?region=${region}`));
+      return data;
+    }
+    default:
+      return [];
```
Comment (amirfefer) [inline src/API/index.js:83]: `react-query` function needs to be a promise, I double-checked, and returning anything in an `async` block, returns a promise with the status `resolved`, so it should work as expected :)

**Review by amirfefer — COMMENTED** [review]
Thanks, @avitova, looks great!
For marking the select warning validation, you can propagate a `validate` prop with a `warning` value to the source select component.

I guess in some cases the list of missing permissions might be long for a helperText, an `ExpandedSection` like @MariSvirik  suggested sounds like a nice solution, also adding some further information on how to add the permissions in the expanded section. You can combine the ExpandedSection  with the helperText component by using the `toggleContent` prop., as you can see in [this](https://codesandbox.io/s/ry7zxv?file=/index.tsx:557-570) live demo

**Comment by avitova** [comment]
image
Thank you both @MariSvirik and @amirfefer, for your comments. I have tried to do the permission check more general, even though it is implemented only for AWS for now.
I thought we could also have a ToolTip for missing permissions, but the list could get really long, so maybe it is not the best idea. The expandable section looks like a good idea. I am a bit worried about the scrollbar though. Let me know what you think.

**Comment by MariSvirik** [comment]
@avitova 
I talked to other designers and we decided on an [inline expandable alert](https://www.patternfly.org/components/alert/html-demos/inline-alert-in-horizontal-form/).
Patternfly doesn't use an expandable plain alert component - as displayed on your screenshot. When there is a problem in the field, use [this](https://www.patternfly.org/components/alert/html-demos/inline-alert-in-horizontal-form/) or [this](https://www.patternfly.org/components/forms/form-control#select) for select. (An inline alert may or may not be used there)

my proposal:
the message is general. When expanded, users can see the list of permissions and where to check it. Would it be possible to add a link where they can check those permissions?
Screenshot 2023-08-24 at 14 37 57
Screenshot 2023-08-24 at 14 40 45

**Comment by avitova** [comment]
image
Nice! We are getting closer and closer. Thank you for your patience. :heart:

**Comment by avitova** [comment]
chrome-capture-2023-7-25 (5)
After conversations with Maria, I have incorporated her input into the changes made.

**Comment by lzap** [comment]
Idea: Implement some limit, if list of missing permission is longer than let’s say 5 you can generate: perm1, perm2, perm3 and 2 more.

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @avitova! It looks great! 
I left a few inline comments, mostly nitpicks, and also please verify that the form validation is not broken due to the `default` addition to the region field.

File: src/Components/ProvisioningWizard/steps/AccountCustomizations/aws.js [hunk 5]
```diff
@@ -34,18 +45,52 @@ const AccountCustomizationsAWS = ({ setStepValidated, image }) => {
...
-        fieldId="aws-select-source"
-      >
+      {(missingPermissions || []).length != 0 && (
+        <FormAlert>
+          <Alert isExpandable isInline variant="warning" title={`Launch might fail due to missing permissions.`}>
+            <>
+              <p>
+                Check if <a href="https://console.aws.amazon.com/iam/">policies</a> in your {image.provider} account for the selected region are set
+                as{' '}
+                <a href="https://github.com/RHEnVision/provisioning-backend/blob/main/docs/configure-amazon-role.md#service-account-policy">
```
[the remark this answers is not in the document]
Comment (ezr-ondrej) [reply src/Components/ProvisioningWizard/steps/AccountCustomizations/aws.js:83]: Could we ... in theory... have the permission list extracted to a file in the backend and load it in code here on the build time - or figure out how to load it to documentation on build time?

We can also play with it as next step tho. I'd either link Sources directly with "Make sure you're following the steps while creating source", or link the official sources documentation with similar message :shrug:

[the remark this answers is not in the document]
Comment (ezr-ondrej) [reply src/Components/ProvisioningWizard/steps/AccountCustomizations/aws.js:83]: Linking to developer docs is not terrible, just not profesional, but given we are explaining the process quite deeply there I'm mostly worried customers might get confused by all the info there :)

[the remark this answers is not in the document]
Comment (avitova) [reply src/Components/ProvisioningWizard/steps/AccountCustomizations/aws.js:83]: I am not sure how we'd want to show the long list here. The link seems like the only sensible option, but I am open if you see a way to show it. We could have one more expandable section with the required permissions if that makes sense.

[the remark this answers is not in the document]
Comment (avitova) [reply src/Components/ProvisioningWizard/steps/AccountCustomizations/aws.js:83]: Do you suggest we hardcode the list of permissions to the official docs for now? Yeah, we could do that, and add it to the [https://github.com/RHEnVision/provisioning-backend/blob/main/CONTRIBUTING.md#basic-guidelines-for-code-contributions](list of places to change).
At the moment, we need to load the permission list to:
- Go file
- Javascript file
- yaml
- markdown

And we can try to eliminate some of these. :thinking:

[the remark this answers is not in the document]
Comment (ezr-ondrej) [reply src/Components/ProvisioningWizard/steps/AccountCustomizations/aws.js:83]: I'm suggesting to split the IAM policy: https://github.com/RHEnVision/provisioning-backend/pull/682 and the include it on build time here

[the remark this answers is not in the document]
Comment (ezr-ondrej) [reply src/Components/ProvisioningWizard/steps/AccountCustomizations/aws.js:83]: Contributing guide is perfect first step for sure :)
